nmpolicy/internal/resolver/walk.go

```diff
@@ -18,49 +18,51 @@ package resolver
 
 import (
 	"fmt"
-	"strconv"
 
 	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
 )
 
-type captureEntryNameAndSteps struct {
-	captureEntryName string
-	steps            ast.VariadicOperator
+func walk(inputState map[string]interface{}, pathSteps ast.VariadicOperator) (interface{}, error) {
+	visitResult, err := visitState(newPath(pathSteps), inputState, &walkOpVisitor{})
+	if err != nil {
+		return nil, fmt.Errorf("failed walking path: %w", err)
+	}
+
+	return visitResult, nil
+}
+
+type walkOpVisitor struct{}
+
+func (*walkOpVisitor) visitLastMap(p path, mapToAccess map[string]interface{}) (interface{}, error) {
+	v, ok := mapToAccess[*p.currentStep.Identity]
+	if !ok {
+		return nil, pathError(p.currentStep, "step not found at map state '%+v'", mapToAccess)
+	}
+	return v, nil
+}
+
+func (*walkOpVisitor) visitLastSlice(p path, sliceToAccess []interface{}) (interface{}, error) {
+	if len(sliceToAccess) <= *p.currentStep.Number {
+		return nil, pathError(p.currentStep, "step not found at slice state '%+v'", sliceToAccess)
+	}
+	return sliceToAccess[*p.currentStep.Number], nil
+}
+
+func (w *walkOpVisitor) visitSlice(p path, sliceToVisit []interface{}) (interface{}, error) {
+	if p.currentStep.Number == nil {
+		return nil, pathError(p.currentStep, "unexpected non numeric step for slice state '%+v'", sliceToVisit)
+	}
+	interfaceToVisit, err := w.visitLastSlice(p, sliceToVisit)
```

Comment:
It is not very readable calling `visitSlice` even if it i not really the last slice.
Maybe a new method can be introduced `getSliceEntry` to be used here and by `visitLastSlice`.

Reply:
This is because we are visiting the last step, not the lastSlice, what do you think about the following visitor interface names ?
- visitStepWithSlice
- visitStepWithMap
- visitLastStepWithSlice
- visitLastStepWithMap

Reply:
We invoke it on lastStep not lastSlice that's why it's missleading.

Reply:
It's invoked only when `p.hasMoreSteps` is false so it's invoked only on the last step.

Reply:
Missunderstood fixing it.